### PR TITLE
Calculate situation-level GNN test accuracy

### LINKED ISSUE #1212: Calculate GNN's situation-level test accuracy
To make it easier to relate ADAM accuracy to GNN accuracy it would be nice to calculate the GNN's situation-level test accuracy as opposed to only calculating the object-level test accuracy. It would be good to have those numbers for the report, especially to explain the one weird color segmentation result where ADAM test accuracy is higher than the GNN's object-level test accuracy (but lower than the GNN's situation-level test accuracy).

### PR DESCRIPTION
**Pull request by spigo900**
Closes #1212.

### CODE REVIEW

File: adam_preprocessing/utils.py [hunk 1]
```diff
@@ -231,6 +231,52 @@ def accuracy(output, target, topk=(1,)):
...
+    # jac: Shouldn't be necessary, given we already forced pred to be .type_as(target), but it was
+    # in Sheng's accuracy code and I didn't care to experiment and find out.
+    target = target.type_as(pred)
+
+    n_situations = 0
+    n_correct = 0
+    for situation, object_indices in situation_number_to_object_indices.items():
+        n_situations += 1
+        if object_indices:
+            if max(object_indices) < pred.shape[0]:
```
Comment (boyleconnor) [inline adam_preprocessing/utils.py:266]: I'm not sure why this case is singled out as one to watch out for and provide a clear error message for

File: adam_preprocessing/utils.py [hunk 2]
```diff
@@ -231,6 +231,52 @@ def accuracy(output, target, topk=(1,)):
     return res
 
 
+def get_situation_accuracy(
+    output: torch.Tensor,
+    target: torch.Tensor,
+    situation_number_to_object_indices: Mapping[int, Sequence[int]],
+) -> float:
+    """
+    Computes the situation-wise acurracy@k for the specified values of k.
```
Comment (boyleconnor) [inline adam_preprocessing/utils.py:240]: This docstring makes it sound like the function returns the accuracy of a particular situation, i.e. `situation_k`. I would not mention the variable `k` in this docstring and instead just say that the function calculates situation-wise accuracy across all situations

File: adam_preprocessing/utils.py [hunk 3]
```diff
@@ -231,6 +231,52 @@ def accuracy(output, target, topk=(1,)):
...
+
+    n_situations = 0
+    n_correct = 0
+    for situation, object_indices in situation_number_to_object_indices.items():
+        n_situations += 1
+        if object_indices:
+            if max(object_indices) < pred.shape[0]:
+                situation_preds = pred[object_indices]
+                n_correct += torch.any(
+                    situation_preds == target[object_indices].expand_as(situation_preds)
```
Comment (boyleconnor) [inline adam_preprocessing/utils.py:269]: Do I understand correctly that this considers an entire situation "correctly predicted" if any object in the situation is predicted with the correct label? If so, I think that should be stated in the docstring.

File: adam_preprocessing/shape_stroke_graph_learner.py [hunk 4]
```diff
@@ -161,16 +163,54 @@ def main():
                 topk=(1,),
             )[0]
         )
+        situation_test_acc = get_situation_accuracy(
+            nn.LogSoftmax(dim=1)(
```
Comment (boyleconnor) [inline adam_preprocessing/shape_stroke_graph_learner.py:167]: I don't think there's any reason to take the softmax before inputting to `get_situation_accuracy()`. It should work correctly directly on the logits

File: adam_preprocessing/shape_stroke_graph_learner.py [hunk 5]
```diff
@@ -161,16 +163,54 @@ def main():
...
+                        test_edge_features,
+                    )
+                )
+            ).data,
+            test_label.data,
+            topk=(1,),
+        )[0]
+    )
+    situation_test_acc = get_situation_accuracy(
+        nn.LogSoftmax(dim=1)(
```
Comment (boyleconnor) [inline adam_preprocessing/shape_stroke_graph_learner.py:203]: Same as my above comment on softmax

**Review by boyleconnor — CHANGES_REQUESTED** [review]

File: adam_preprocessing/shape_stroke_graph_learner.py (hunk 4 above)
Comment (spigo900) [reply adam_preprocessing/shape_stroke_graph_learner.py:167]: Fair enough.

File: adam_preprocessing/utils.py (hunk 3 above)
Comment (spigo900) [reply adam_preprocessing/utils.py:269]: Correct.

File: adam_preprocessing/utils.py (hunk 2 above)
Comment (spigo900) [reply adam_preprocessing/utils.py:240]: My bad, that came from the original accuracy function's docstring. I'll remove it.

File: adam_preprocessing/utils.py (hunk 1 above)
Comment (spigo900) [reply adam_preprocessing/utils.py:266]: I suppose it's not strictly necessary here.

File: adam_preprocessing/shape_stroke_graph_learner.py (hunk 4 above)
Comment (spigo900) [reply adam_preprocessing/shape_stroke_graph_learner.py]: Removed the softmax.

File: adam_preprocessing/shape_stroke_graph_learner.py (hunk 5 above)
Comment (spigo900) [reply adam_preprocessing/shape_stroke_graph_learner.py]: Removed the softmax.

File: adam_preprocessing/utils.py (hunk 3 above)
Comment (spigo900) [reply adam_preprocessing/utils.py]: Added this to the docstring.

File: adam_preprocessing/utils.py (hunk 2 above)
Comment (spigo900) [reply adam_preprocessing/utils.py]: Removed the `@k` part from the docstring.

File: adam_preprocessing/utils.py (hunk 1 above)
Comment (spigo900) [reply adam_preprocessing/utils.py]: Removed the special error message.

**Review by lichtefeld — APPROVED** [review]
looks good to me

**Review by boyleconnor — APPROVED** [review]
